### ISSUE TITLE
Error `Assertion failed: Nonzero carryout of nonoverlapping vector add!` on x86 when compiled with GCC and LTO enabled

### ISSUE THREAD

**Issue by tdulcet** [primesearch/Mlucas#10]
Mlucas cannot be compiled on x86 with GCC and Link Time Optimization (LTO) enabled, as it triggers a runtime error:
```
INFO: Maximum recommended exponent for FFT length (2048 Kdbl) = 39606917; p[ = 39003229]/pmax_rec = 0.9847580159.
Initial DWT-multipliers chain length = [short] in carry step.
M39003229: using FFT length 2048K = 2097152 8-byte floats, initial residue shift count = 31137256
This gives an average   18.598188877105713 bits per digit
Using complex FFT radices      1024        32        32
mers_mod_square: Init threadpool of 1 threads
Using 1 threads in carry step
ERROR: at line 457 of file ../src/mi64.c
Assertion failed: Nonzero carryout of nonoverlapping vector add!
```
Here is the assert: https://github.com/primesearch/Mlucas/blob/d2dad70871af597919b51377251fbfcf55329d6c/src/mi64.c#L457
The issue seems to affect all FFT lengths. If LTO could be enabled, it could provide a significant performance improvement.

**Edit**: We got [a report](https://www.mersenneforum.org/showpost.php?p=654167&postcount=92) that LTO works on 32-bit ARM.

**Comment by gbvalor** [comment]
On apple silicon  M3 Pro with armv8 (64-bits) and compiled with **-flto=auto** , **Mlucas -s a**  completed selftest without problems. 

OTOH, timings are the same that in no-lto case. No gain using LTO in my tests.

**Comment by ldesnogu** [comment]
> OTOH, timings are the same that in no-lto case. No gain using LTO in my tests.

Depending on what you're running, most of the executed code can be in hand-written assembly loops so the compiler won't be able to improve run time a lot.

**Comment by tdulcet** [comment]
> On apple silicon M3 Pro with armv8 (64-bits) and compiled with **-flto=auto** , **Mlucas -s a** completed selftest without problems.

Interesting, thanks for the information. I was able to confirm that LTO works on my 64-bit ARM system as well. We got another report that LTO works on 32-bit ARM, so the bug must only affect x86 systems.

> No gain using LTO in my tests.

Hmm, based on previous reports from users enabling LTO with Prime95/MPrime, I would expect at least a 5% speedup. Considering that Mlucas is already up to 50% slower than Prime95/MPrime, it needs all the speedups that it can get. May I ask how you enabled LTO specifically?

**Comment by gbvalor** [comment]
@tdulcet , I enabled LTO just uncommenting the end of line 373 in makemake.sh file. I also saw that '-flto=auto' was in build.log.  I just ran a selftest with the resulting LTO binary.  

I will compare carefully the timings with and without LTO. This was a first impression.

**Comment by gbvalor** [comment]
And I forgot to tell that I also built with LTO using gcc (in macosx was just clang), and gcc-13 (genuine gcc-13) installed with brew. No problems on both tries.

**Comment by tdulcet** [comment]
> I enabled LTO just uncommenting the end of line 373 in makemake.sh file.

Thanks for the additional information. If you previously built Mlucas (without LTO), be sure to run `make clean` in the `obj` directory to remove all of the object files before recompiling with LTO enabled. For example:
```bash
cd obj
make clean
rm Mlucas
cd ..
bash makemake.sh use_hwloc
```
When running the self-tests, I would recommend running (add the `-cpu` or `-core` arguments as needed): `./Mlucas -s a -iters 10000` for more accurate timings and then compare the resulting ms/iter speeds in the `mlucas.cfg` file.

**Comment by ldesnogu** [comment]
This doesn't look like a bug:

			mi64_div(c_uint64_ptr, &itmp64, j+1,1, arrtmp,&rmodb);	ASSERT(HERE, rmodb == 0ull,"After short-div, R != 0 (mod B)");

int mi64_div(const uint64 x[], const uint64 y[], uint32 lenX, uint32 lenY, uint64 q[], uint64 r[])
	/* And now find the actual lengths of the divide operands and use those for the computation: */
	xlen = mi64_getlen(x, lenX);
	ylen = mi64_getlen(y, lenY);

       if((xlen < ylen) || ((xlen == ylen) && mi64_cmpult(x, y, max_len)) ) {
		/* If no modular reduction needed and X and R not the same array, set R = X: */
		if(r != 0x0 && r != x) {
			mi64_set_eq(r, x, lenX);
		}
		if(q) mi64_clear(q,lenX);
		return mi64_iszero(x, lenX);	// For x < y, y divides x iff x = 0
	}

So xlen will be <= ylen which is 1.

**Comment by tdulcet** [comment]
@ldesnogu - Thanks for looking into that warning.

I was able to trigger this assertion with GCC on Linux without enabling LTO. It happened when I was attempting to fix the 10 `discarded-qualifiers` warnings (see commit: 028b8d0bc27ae093186bb34ff95ab260d4814aca), so it might be related to some variable that is labeled as `const` when it should not be. It seem the compiler does not like creating pointers to constant variables.

**Comment by tdulcet** [comment]
From further testing, I have narrowed to scope of the bug to x86-64 systems when compiled with GCC. It seems to be triggered by compiler optimizations, like those facilitated by enabling LTO or my proposed changes to Ernst's `ASSERT` function. I moved those changes to the `ASSERT` function from #22 to a separate PR (#24), where I also enabled LTO by default. If anyone wants to help test this, the changes are on the [`asserts` branch](https://github.com/primesearch/Mlucas/tree/asserts). Obviously, that PR and any other changes that could enable further compiler optimizations are now blocked on fixing this issue.

**Comment by KarlLudwig3485** [comment]
I've tested builds with different combinations of optimisation flags, and GCC with LTO works with -O1, but not 2 or 3. 

|Compiler|Flags|Result|
|-|-|-|
|GCC 14.2.1|-O3|✔️|
|GCC 14.2.1|-O3 -flto=auto|✖|
|GCC 14.2.1|-O2 -flto=auto|✖|
|GCC 14.2.1|-O1 -flto=auto|✔️|
|Clang 18.1.8|-O3|✔️|
|Clang 18.1.8|-O3 -flto=auto|✔️|

**Comment by KarlLudwig3485** [comment]
I've tested some builds with different combinations of optimisation flags, and GCC with LTO works with -O1, but not 2 or 3.
After checking the GCC documentation and manually adding every flag enabled by -O2 one by one, I've determined that the offending flag is `-finline-small-functions`. Disabling it with `-fno-inline-small-functions` results in a working build. That's about as far as my limited knowledge can take me, so I'm hoping this goes some way to diagnosing the actual issue.

|Compiler|Flags|Result|
|-|-|-|
|GCC 14.2.1|-O3|✔|
|GCC 14.2.1|-O3 -fno-inline-small-functions -flto=auto|✔|
|GCC 14.2.1|-O3 -flto=auto|✖|
|GCC 14.2.1|-O2 -flto=auto|✖|
|GCC 14.2.1|-O1 -flto=auto|✔|
|Clang 18.1.8|-O3|✔|
|Clang 18.1.8|-O3 -flto=auto|✔|

Target is Linux x86_64 AVX2, built from `main` at 4db9fbc42ba00298c123623549bd36dd47573d45.

**Comment by tdulcet** [comment]
> That's about as far as my limited knowledge can take me, so I'm hoping this goes some way to diagnosing the actual issue.

Thanks, that sounds like a lot of work, but is very helpful. With this information and from looking at the code again, the issue is almost certainly with that `mi64_add()` function that is called immediately before the assert statement. On ARM where LTO works as expected, this function has a pure C implementation, but on x86 it is written in inline assembly. The LTO when combined with that `inline-small-functions` option is causing the entire function to be inlined, which is obviously breaking some assumption made by the assembly code:
https://github.com/primesearch/Mlucas/blob/d2dad70871af597919b51377251fbfcf55329d6c/src/mi64.c#L2241-L2471
We would need someone who has experience with inline x86 assembly to debug this. In the meantime, we could switch this to use the pure C implementation on x86, but the performance degradation from doing that may be greater than what would be gained from enabling LTO.

**Comment by KarlLudwig3485** [comment]
Having learned a bit about inline assembly, how to properly use a debugger, and added some test code to compare the results of `mi64_add()` with `mi64_add_ref()`, I determined that the output of `mi64_add()` is correct, except for the carry-out.

The `__cy` operand in the inline assembly was incorrectly set as write-only (`=`), causing the `cy = 0` init to be optimized out. Setting `__cy` to read/write (`+`) fixes this. Interestingly the `cy = 0` init failing was a known issue in 2016 according to the comments above it, but the cause wasn't known.

I've made a pull request. :)
